### PR TITLE
Add a CLI tool for code generation

### PR DESCRIPTION
**Pull request by matiasbeckerle**
Hi folks. One of the @mravinale's ideas is to have a tiny CLI tool bundled in the seed project.

In this PR you will see the first draft of that CLI. Based on Yeoman, I added the necessary scaffolding stuff to start working out in new generators. I've included one in order to play a little bit and I also believe is useful.

Installation and usage is also documented (please read in the diff).

For demoing purposes here is a video: https://www.screencast.com/t/JGNiwoueQ5

Please feel free to comment about anything. Even about the component template. Shoot!

@mravinale @juampick

### CODE REVIEW

File: cli/generators/component/templates/component.js
Comment (AlphaGit) [inline cli/generators/component/templates/component.js]: Is `PropTypes` being used at all?

File: cli/generators/component/templates/component.js
Comment (AlphaGit) [inline cli/generators/component/templates/component.js]: Why is this assigned after the fact and not generated inside the class? Is this a class-level assignment?

File: cli/generators/component/templates/component.js
Comment (matiasbeckerle) [inline cli/generators/component/templates/component.js]: Good question. As you noticed is not being used in the template but: if we are going to prefer adding `propTypes` and `defaultProps` because 83% of the components uses it then `PropTypes` is going to be used and it makes sense to add it.

But that question is why I don't like code generation too much hehe. Is a gray middle-ground. What do you think?

File: cli/generators/component/templates/component.js
Comment (matiasbeckerle) [inline cli/generators/component/templates/component.js]: ...water...

Unfortunately I don't know the insights of this yet. I'm discovering it step by step. It work like a class property and I bet is a convention in the React community to put it outside. I can be wrong anyways.

File: cli/generators/component/templates/component.js
Comment (AlphaGit) [inline cli/generators/component/templates/component.js]: If most of them will actually use them, I think it's fine. It's going to save you more time if it has it and you (sometimes) need to remove it than if you don't have it and almost every time you need to add it.

File: cli/.editorconfig
Comment (juampick) [inline cli/.editorconfig]: What do you think leaving this `.editorconfig` file equal to the main `.editorconfig` of the root on *mern-seed*? Or do we need also this? Thanks @matiasbeckerle

File: cli/README.md
Comment (juampick) [inline cli/README.md]: What do you think @matiasbeckerle to update this command or an optional parameter to also create a folder with the same name? 
ie: `yo mern:component Sidebar -f` and will create the component Sidebar inside: */app/components/sidebar/Sidebar.js*

File: cli/generators/component/templates/component.js
Comment (juampick) [inline cli/generators/component/templates/component.js]: As we discussed @matiasbeckerle please change it to have the propTypes and defaultProps below. Thanks!

File: cli/.editorconfig
Comment (matiasbeckerle) [inline cli/.editorconfig]: Good point. Here are my thoughts about the matter:

I added this `.editorconfig` because I wanted to treat the cli as "kind of independent project with his own rules" but now that you pointed out I believe it is ok if I remove it and we start using the root one.

My only concern and this is one that I will not easily grant/giveaway is that I love the two spaces. LOL. Yes. I remember a discussion one or two years ago about the matter because the web is going in that front instead of the 4 spaces (web dev, not other like .NET where the 4 spaces rocks!). I remember that I forced myself to push the web standards and start using the 2 spaces. An eventually really loved it.

So, recap. My suggestion is to have only one `.editorconfig` at root and start using the 2 spaces for indentation. If I have greenlight from the team I will leave this `.editorconfig` as is and create a separate ticket to do those changes.

Thoughts, opinions, ideas? @juampick @mravinale @AlphaGit

File: cli/README.md
Comment (matiasbeckerle) [inline cli/README.md]: You know what? While I was doing this whole thing I thought about it because _for sure_ is something that will be used. Let me see what can I do to achieve it. Thanks!

Comment (juampick) [inline cli/README.md]: Hi @matiasbeckerle ... I was thinking while sleeping and I think the interesting thing that we can do here is to have 3 different commands (or 3 differents optional parameters - I leave that up to you) to create the 3 different type of components you can have in React-Redux project then.
For example: (I'll not create the templates but simple examples)
1- `React Class Component` (the one you've created)
```
import React, {PropTypes} from 'react';

class ReactClassComponent extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <div>
                ---Autogenerated <%= name %>---
            </div>
        );
    }
}

ReactClassComponent.defaultProps = {

};

ReactClassComponent.propTypes = {

};

export default ReactClassComponent
```

2- `React Stateless Component` 
```
import React, {PropTypes} from 'react';

const ReactStatelessComponent = (props) => {
    return (
        <div></div>
    );
};

ReactStatelessComponent.defaultProps = {
    
};

ReactStatelessComponent.propTypes = {
    
};

export default ReactStatelessComponent;
```

3- `React-Redux Container Component`
```
import React, {PropTypes} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';

class ReactReduxContainerComponent extends React.Component {
    constructor(props, context) {
        super(props, context);
    }

    render() {
        return (
            <div></div>
        );
    }
}

ReactReduxContainerComponent.defaultProps = {
   
};

ReactReduxContainerComponent.propTypes = {
   
};

function mapStatesToProps(state, ownProps) {
    return {
        state: state
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(actions, dispatch)
    }
}

export default connect(mapStatesToProps, mapDispatchToProps)(ReactReduxContainerComponent);
```

Do you like the idea @matiasbeckerle ?

Comment (matiasbeckerle) [inline cli/README.md]: Sound like an awesome plan. I would like to be really aware of the insights of each one. Thanks. Will implement and let you know.

File: cli/.editorconfig
Comment (juampick) [inline cli/.editorconfig]: I understand your thoughts @matiasbeckerle. I don't see a bad thing to have it, as a separate project the generator side.

I'm not sure that there's a `good way` or `the way`, (It's a really old discussion but I think it's not an important thing to spend time) but if I need to say something I may prefer the 4 spaces. More clear and neat (for me obviously). 🖖

File: cli/README.md
Comment (juampick) [inline cli/README.md]: 👍 @matiasbeckerle thanks! 
No problem, I can explain the insights of each one when you want! thanks 💯

Comment (matiasbeckerle) [inline cli/README.md]: Done the folder thing in https://github.com/MakingSense/mern-seed/pull/10/commits/3c2218b5bbf4dd16f342c24fa0c7d3d95a1b7d44. See the docs!

File: cli/generators/component/templates/component-container.js
Comment (matiasbeckerle) [inline cli/generators/component/templates/component-container.js]: Ideas on a better way to handle this? I couldn't think a quick solution. @juampick

**Comment by matiasbeckerle** [comment]
Please take another look @juampick. Specially this last commit https://github.com/MakingSense/mern-seed/pull/10/commits/f897e58f787f1d470ea9fd8e4e4d5d114b332393 where I implemented what we talked about.

Side though: optional parameters as the name represents, are optional. With my changes if anyone uses --stateless --container it will only take into account one of them. Is like not optional. I have different ways to solve it but I wanted to use the simplicity of using flags, because are nice shortcuts. If you think I should use a user dialog in order to pick a type let me know and I can do the change but it will be an extra step the dev will need to fill.

File: cli/.editorconfig
Comment (AlphaGit) [inline cli/.editorconfig]: I agree with the two-spaces approach, but I wouldn't make this a **root** `.editorconfig`. I can imagine a project where both frontend and backend standards are at play and they contradict each other. In the .NET world, for example, it is pretty standard to have 4 spaces or tabs.

So, I think we should keep the `.editorconfig` with the 2 spaces, but not make it a root. As such, it would only affect what is generated with this seed (which is intended) and would not clash with any other standard in the grand-scheme of other projects (which I believe is desirable).

Comment (matiasbeckerle) [inline cli/.editorconfig]: Understand. Then I will leave this `.editorconfig` as is and update the root one in order to use the 2 spaces too (all of this in another PR). In this particular project both are web related but at least opens the door to do something else.

File: cli/generators/component/templates/component-container.js
Comment (juampick) [inline cli/generators/component/templates/component-container.js]: Yes, i think it's a good approach for now ;)

File: cli/README.md
Comment (juampick) [inline cli/README.md]: Awesome readme @matiasbeckerle 💯

**Comment by juampick** [comment]
I understand @matiasbeckerle your concern but I think it's nice for now.. as a 1st approach! congratz

File: cli/README.md
Comment (matiasbeckerle) [inline cli/README.md]: Thank you!!!

**Review by juampick — APPROVED** [review]
Looking ok.

**Pull request merged**
